common/blobstore/s3store/client.go

```diff
@@ -0,0 +1,311 @@
+	if err != nil {
+		if err == ErrReadFile {
+			return nil, blobstore.ErrBlobNotExists
+		}
+		return nil, err
+	}
+	tags, err := s3gettags(ctx, c.s3cli, bucket, key.String())
+	if err != nil {
+		return nil, err
+	}
+	return blob.NewBlob(b, tags), nil
+}
+
+func (c *client) GetTags(ctx context.Context, bucket string, key blob.Key) (map[string]string, error) {
+
+	tags, err := s3gettags(ctx, c.s3cli, bucket, key.String())
+	if err != nil {
+		return nil, err
+	}
+	return tags, nil
+}
+
+func (c *client) Exists(ctx context.Context, bucket string, key blob.Key) (bool, error) {
+
+	_, err := c.s3cli.HeadObjectWithContext(ctx, &s3.HeadObjectInput{
+		Bucket: aws.String(bucket),
+		Key:    aws.String(key.String()),
+	})
+	if err != nil {
+		aerr, ok := err.(awserr.Error)
+		if ok && aerr.Code() == s3.ErrCodeNoSuchKey {
+			return false, nil
+		}
+		if ok && aerr.Code() == s3.ErrCodeNoSuchBucket {
+			return false, blobstore.ErrBucketNotExists
+		}
+
+		return false, err
+	}
+	return true, nil
+}
+
+func (c *client) Delete(ctx context.Context, bucket string, key blob.Key) (bool, error) {
+
+	_, err := c.s3cli.DeleteObjectWithContext(ctx, &s3.DeleteObjectInput{
+		Bucket: aws.String(bucket),
+		Key:    aws.String(key.String()),
+	})
+	if err != nil {
+		if aerr, ok := err.(awserr.Error); ok {
+
+			if aerr.Code() == s3.ErrCodeNoSuchBucket {
+				return false, blobstore.ErrBucketNotExists
+			}
+
+			if aerr.Code() == s3.ErrCodeNoSuchKey {
+				return false, nil
+			}
+
+		}
+		return false, err
+
+	}
+	return true, nil
+}
+
+func (c *client) ListByPrefix(ctx context.Context, bucket string, prefix string) ([]blob.Key, error) {
+
+	results, err := c.s3cli.ListObjectsV2WithContext(ctx, &s3.ListObjectsV2Input{
+		Bucket: aws.String(bucket),
+		Prefix: aws.String(prefix),
+	})
+	if err != nil {
+		if aerr, ok := err.(awserr.Error); ok && aerr.Code() == s3.ErrCodeNoSuchBucket {
+			return nil, blobstore.ErrBucketNotExists
+		}
+		return nil, ErrListFiles
+	}
+	var keys = make([]blob.Key, len(results.Contents))
+	for i, v := range results.Contents {
+		key, err := blob.NewKeyFromString(*v.Key)
+		if err != nil {
+			return nil, ErrConstructKey
+		}
+		keys[i] = key
+	}
+	return keys, nil
+}
+
+func (c *client) BucketMetadata(ctx context.Context, bucket string) (*blobstore.BucketMetadataResponse, error) {
+
+	results, err := c.s3cli.GetBucketAclWithContext(ctx, &s3.GetBucketAclInput{
+		Bucket: aws.String(bucket),
+	})
+	if err != nil {
+		if aerr, ok := err.(awserr.Error); ok && aerr.Code() == s3.ErrCodeNoSuchBucket {
+			return nil, blobstore.ErrBucketNotExists
+		}
+		return nil, err
+	}
+
+	lifecycleResults, err := c.s3cli.GetBucketLifecycleWithContext(ctx, &s3.GetBucketLifecycleInput{
+		Bucket: aws.String(bucket),
+	})
+	if err != nil {
+		if aerr, ok := err.(awserr.Error); ok && aerr.Code() == s3.ErrCodeNoSuchBucket {
+			return nil, blobstore.ErrBucketNotExists
+		}
+		return nil, err
+	}
+	var retentionDays = math.MaxInt64
```

Comment:
We use the value of 0 to indicate unbounded retention

```
	if resp.Configuration.ArchivalRetentionPeriodInDays != nil {
		formatStr = formatStr + "ArchivalRetentionInDays: %v\n"
		archivalRetentionDays := resp.Configuration.GetArchivalRetentionPeriodInDays()
		if archivalRetentionDays == 0 {
			descValues = append(descValues, "unlimited")
		} else {
			descValues = append(descValues, archivalRetentionDays)
		}
	}
```

Reply:
Fixed with the caveat that the bucket can have multiple retention rules. I also checked that the rules are enabled. It'll return the largest expiration it can find now